lxd/instance/drivers/driver_lxc.go

```diff
@@ -3646,13 +3651,16 @@ func (c *lxc) Rename(newName string) error {
 	}
 
 	for _, backup := range backups {
-		backupName := strings.Split(backup.Name(), "/")[1]
+		oldName := backup.Name()
+		backupName := strings.Split(oldName, "/")[1]
 		newName := fmt.Sprintf("%s/%s", newName, backupName)
 
 		err = backup.Rename(newName)
 		if err != nil {
 			return err
 		}
+
+		revert.Add(func() { backup.Rename(oldName) })
```

Comment:
@monstermunchkin be careful with this pattern: adding a revert step inside a for loop using the iterator variable.

I've been caught out by that myself.

Try this simple example:

```go
package main

import (
	"fmt"

	"github.com/lxc/lxd/lxd/revert"
)

func main() {
	revert := revert.New()
	defer revert.Fail()

	things := []string{"one", "two", "three"}

	for _, thing := range things {
		revert.Add(func() { fmt.Println(thing) })
	}
}
```

This prints:

```
three
three
three
```

So you'll end up reverting the last snapshot multiple times. This is basically the same issue as using defer in a loop too.

This is because the loop iterator variable is only created once per loop, not once per iteration, and its value is just changed for each loop. Which means when the reverter runs it takes the last value of that variable and passes it into each revert step.

Instead you can either define a local variable inside the loop that copies the iterator value and use that inside the reverter, e.g.

```go
package main

import (
	"fmt"

	"github.com/lxc/lxd/lxd/revert"
)

func main() {
	revert := revert.New()
	defer revert.Fail()

	things := []string{"one", "two", "three"}

	for _, thing := range things {
		name := thing
		revert.Add(func() { fmt.Println(name) })
	}
}
```

Or you can make each iteration its own atomic unit by putting it inside an anonymous function (which is also useful when doing defer to avoid stacking them up until the end of the loop).

```go
package main

import (
	"fmt"

	"github.com/lxc/lxd/lxd/revert"
)

func main() {
	revert := revert.New()
	defer revert.Fail()

	things := []string{"one", "two", "three"}

	for _, thing := range things {
		func(name string) {
			revert.Add(func() { fmt.Println(name) })
		}(thing)
	}
}
```

Reply:
I've just spotted you are using a local variable, so should be fine. Ignore me :)

Reply:
Had you done `revert.Add(func() { backup.Rename(backup.Name()) })` that would have been an issue.

Reply:
@monstermunchkin oh no, it is an issue, because `backup` is always going to reference the last snapshot, so whilst the `oldName` value is correct, you'll be renaming the wrong snapshot.

Reply:
Here's an updated example:

```go
package main

import (
	"fmt"

	"github.com/lxc/lxd/lxd/revert"
)

type thing struct {
	name string
}

func (t *thing) Name() string {
	return t.name
}

func main() {
	revert := revert.New()
	defer revert.Fail()

	things := []thing{thing{"one"}, thing{"two"}, thing{"three"}}

	for _, thing := range things {
		revert.Add(func() { fmt.Println(thing.Name()) })
	}
}
```

prints:

```
three
three
three
```

Reply:
This should be fixed now. Thanks for that :)